### PR TITLE
Direct velocity primitive implementation

### LINKED ISSUE #97: grSim backend - Implement DirectVelocity Primitive
[grSim](https://github.com/RoboCup-SSL/grSim) is the "standard" simulator for the SSL Division, and is the simulator we use.

Because we currently don't/can't simulate our robot's firmware, we need to re-create how we move and control the robots for the simulator. Rather than sending a Primitive command to a robot over radio, which will then be executed by the robot firmware in order to do whatever the command asks, we need to send different commands to the simulator to move the robots.

We already have the ability to send commands to the simulator with #12 . The ability to control a robot's movement is further abstracted by the motion controller developed in #9 . This imitates the bang-bang motion controller on our robots, so that robot's will speed up and slow down more realistically as the move in the simulator.

** Currently blocked by #9  **

The task here is to create a function/routine to implement our DirectVelocity Primitive using this motion-controller abstraction, so this Primitive can be simulated. We need these primitive to work in the simulator so we can actually use the simulator for testing.

The spec for the DirectVelocity Primitive can be found in #86 , and also in the DirectVelocityPrimitive class from #91 

The DirectVelocityPrimitive class will be the input. The exact place this code will go will be revised once #9 is merged, and we will update the ticket.

### PR DESCRIPTION
**Pull request by HaomiaoZ**
### Description

Implement Direct velocity primitive in grSim backend. Calculate desired final position and orientation based on given direct velocity primitive and create a `motion_controller_command` to control the robot in grSim through bang-bang motion controller. Dribbler state is determined from the dribbler speed.

### Testing Done

Testing was done through grSim by visualising the movement of a robot based on direct velocity primitives with different arguments.

### Resolved Issues

resolves #97 

### Length Justification

Length is less than 500 lines

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [ ] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [ ] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Couple comments. Have you tested this in grsim?

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp [hunk 1]
```diff
@@ -29,7 +29,33 @@ void GrsimCommandPrimitiveVisitor::visit(const ChipPrimitive &chip_primitive)
...
 {
-    // TODO: https://github.com/UBC-Thunderbots/Software/issues/97
+    // get current robot position and orientation(angle)
+    Point robot_position    = robot.position();
+    Angle robot_orientation = robot.orientation();
+
+    // create linear velocity vector from direct velocity primitive
+    Vector linear_velocity_in_robot_coordinates =
+        Vector(direct_velocity_primitive.getXVelocity(),
+               direct_velocity_primitive.getXVelocity());
```
[the remark this answers is not in the document]
Comment (HaomiaoZ) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: That is a typo, I will fix it, thanks for catching that

File: src/thunderbots/software/grsim_communication/main.cpp [hunk 2]
```diff
@@ -85,6 +86,9 @@ int main(int argc, char** argv)
         // Spin once to let all necessary callbacks run
         // The callbacks will populate the primitives vector
         ros::spinOnce();
+        // primitives.emplace_back(new MovePrimitive(1, Point(0, 0), Angle::zero(), 0));
+
+        primitives.emplace_back(new DirectVelocityPrimitive(0, 0, 0, 1, 0));
```
[the remark this answers is not in the document]
Comment (HaomiaoZ) [reply src/thunderbots/software/grsim_communication/main.cpp]: ok, I will exclude them from the branch

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp [hunk 3]
```diff
@@ -29,7 +29,33 @@ void GrsimCommandPrimitiveVisitor::visit(const ChipPrimitive &chip_primitive)
...
+    Vector linear_velocity_in_global_coordinates =
+        linear_velocity_in_robot_coordinates.rotate(robot.orientation());
+
+    // TODO final reasonable scalars to multiply linear and angular velocity to get
+    // reasonable speed final destination is the parameter that can control the robot to
+    // move in the direction of velocity vector from current robot position
+    Vector final_destination = linear_velocity_in_global_coordinates + robot_position;
+    // final orientation is the parameter that can control the robot to rotate in the
+    // direction of angular velocity from current robot orientation
+    Angle final_orientation =
```
[the remark this answers is not in the document]
Comment (HaomiaoZ) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp:104]: So Mathew you are suggesting using `final_orientation.clamp()` so that final orientation won't wrap around the robot?

File: src/thunderbots/software/ai/world/team.cpp [hunk 4]
```diff
@@ -76,15 +76,15 @@ void Team::removeExpiredRobots(const std::chrono::steady_clock::time_point times
...
+    /*if (getRobotById(new_goalie_id))
     {
         goalie_id = new_goalie_id;
     }
     else
     {
         throw std::invalid_argument(
             "Error: Assigning the goalie to a robot that is not a member of the team");
-    }
+    }*/
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/team.cpp]: I'm a bit confused, how is this change related?

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp [hunk 5]
```diff
@@ -29,7 +29,35 @@ void GrsimCommandPrimitiveVisitor::visit(const ChipPrimitive &chip_primitive)
...
+        Vector(direct_velocity_primitive.getXVelocity(),
+               direct_velocity_primitive.getXVelocity());
+
+    // transfer velocity into global coordinate by rotating the vector in robot
+    // coordinates by the angle of robot
+    Vector linear_velocity_in_global_coordinates =
+        linear_velocity_in_robot_coordinates.rotate(robot.orientation());
+
+    // TODO find reasonable scalars to multiply linear and angular velocity to get
+    // reasonable speed
```
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: Could we either resolve this TODO or assign an issue to it?

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp:104]: I think it's easier to clamp the angular velocity value you're adding to the robot orientation to be no more than +/- pi/2

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (HaomiaoZ) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp:104]: That make more sense, I will clamp angular velocity  to +/- pi/2 for now then

File: src/thunderbots/software/ai/world/team.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/team.cpp]: This exception is being unexpectedly thrown when running the AI, so this is a temporary fix. A proper fix will need to be investigated/implemented separately. I have a reminder to create a ticket for this.

Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/team.cpp]: Oooh just realized this causes test failures. Going to prioritize a proper fix for this ASAP that should be merged first.

Comment (garethellis0) [reply src/thunderbots/software/ai/world/team.cpp]: Wut, scary.

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp (hunk 5 above)
Comment (HaomiaoZ) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: Currently I'm not sure how to get the reasonable scalar.

Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: @MathewMacDougall?

Comment (MathewMacDougall) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: I just made a new ticket. Link this TODO to #286.

@garethellis0 The problem is essentially that we want to directly set/request the robot have a certain velocity, and can't do that with the current motion controller interface. It's very hard/impossible to set the requested "final state" of the robot such that the robot ends up with the desired velocities.

**Comment by MathewMacDougall** [comment]
The issue with the `assignGoalie` function in the `Team` class throwing exceptions has been temporarily fixed in #288. Make sure to update your branch.

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: Ah, unfortunate. How are we testing that this is working correctly if the motion controller interface isn't compatible with it then?

Comment (MathewMacDougall) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: Visual inspection in grsim that it's close enough for now.

Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by matthewberends — APPROVED** [review]
LGTM :shipit:

**Review by MathewMacDougall — APPROVED** [review]
